src/Raven.Server/Documents/Sharding/Documents/ShardedSubscriptionContext.cs

```diff
@@ -0,0 +1,183 @@
+        {
+            _shardedContext = shardedContext;
+            _serverStore = serverStore;
+        }
+
+        public override string DatabaseName => _shardedContext.DatabaseName;
+
+        public override SubscriptionConnectionState OpenSubscription(SubscriptionConnectionBase connection)
+        {
+            var subscriptionState = _subscriptionConnectionStates.GetOrAdd(connection.SubscriptionId,
+                subscriptionId => new SubscriptionConnectionState(subscriptionId, connection.Options.SubscriptionName, this));
+
+            return subscriptionState;
+        }
+
+        public async Task<bool> DropSubscriptionConnectionAndPropagateToShards(long subscriptionId, SubscriptionException ex)
+        {
+            if (_subscriptionConnectionStates.TryGetValue(subscriptionId, out SubscriptionConnectionState subscriptionConnectionState) == false)
+                return false;
+
+            // drop on shards
+            var disposables = new List<IDisposable>();
+            try
+            {
+                var tasks = new List<Task>();
+                var cmds = new List<DropSubscriptionConnectionCommand>();
+                foreach (var re in _shardedContext.RequestExecutors)
+                {
+                    disposables.Add(_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext ctx));
+                    var cmd = new DropSubscriptionConnectionCommand(subscriptionConnectionState.SubscriptionName);
+                    cmds.Add(cmd);
+                    tasks.Add(re.ExecuteAsync(cmd, ctx));
+                }
+
+                await Task.WhenAll(tasks);
+            }
+            finally
+            {
+                disposables.ForEach(x => x.Dispose());
+            }
+
+            var subscriptionConnection = subscriptionConnectionState.Connection;
+
+            if (subscriptionConnection != null)
+            {
+                // wait for propogate from the server workers to sharded connection
+                if (subscriptionConnection.CancellationTokenSource.Token.WaitHandle.WaitOne(15000) == false)
+                {
+                    try
+                    {
+                        subscriptionConnection.CancellationTokenSource.Cancel();
+                    }
+                    catch
+                    {
+                        // ignored
+                    }
+                }
+
+                subscriptionConnectionState.RegisterRejectedConnection(subscriptionConnection, ex);
+                subscriptionConnection.ConnectionException = ex;
+            }
+
+            if (_logger.IsInfoEnabled)
+                _logger.Info($"Sharded subscription with id '{subscriptionId}' and name '{subscriptionConnectionState.SubscriptionName}' connection was dropped. Reason: {ex.Message}");
+
+            return true;
+        }
+
+        public async Task<SubscriptionState> AssertSubscriptionConnectionDetails(long id, string name)
```

Comment:
Do we need to bother with this? Aren't we going to get the same error from the shards directly?